R/linear_pool_sample.R

```diff
@@ -42,17 +41,55 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
     )
     weights_col_name <- "weight"
   }
+  weight_by_cols <- colnames(weights)[colnames(weights) != weights_col_name]
   unique_weights <- unique(weights[[weights_col_name]])
 
-  if (length(unique_weights) != 1 || length(unique_provided_samples) != 1 || !is.null(n_output_samples)) {
+  if (length(unique_weights) != 1 || length(unique_provided_samples) != 1) {
     cli::cli_abort(
       "The requested ensemble calculation doesn't satisfy all conditions:
       1) {.arg model_out_tbl} contains the same number of samples from each component model,
-      2) {.arg weights} are {.val NULL} or equal for every model,
-      3) {.arg n_output_samples} = {.val NULL}"
+      2) {.arg weights} are {.val NULL} or equal for every model"
     )
   }
 
+  if (!is.null(n_output_samples)) {
+    samples_per_combo <- samples_per_combo |>
+      dplyr::left_join(weights, weight_by_cols) |>
+      dplyr::mutate(target_samples = floor(.data[[weights_col_name]] * n_output_samples))
+
+    if (any(samples_per_combo$provided_samples < samples_per_combo$target_samples)) {
+      cli::cli_abort("Requested output samples per compound unit cannot exceed the provided samples per compound unit.")
+    }
+
+    # deal with n_output_samples not divisible evenly among component models
+    actual_output_samples <- samples_per_combo |>
+      dplyr::group_by(dplyr::across(dplyr::all_of(task_id_cols))) |>
+      dplyr::summarize(target_samples = sum(.data[["target_samples"]])) |>
+      dplyr::pull(.data[["target_samples"]]) |>
+      unique()
+    remainder_samples <- n_output_samples - actual_output_samples
+    models_to_resample <- sample(x = unique(samples_per_combo$model_id), size = remainder_samples)
+    samples_per_combo <- samples_per_combo |>
+      dplyr::mutate(target_samples = ifelse(
+        model_id %in% models_to_resample, .data[["target_samples"]] + 1, .data[["target_samples"]]
+      ))
+
+    split_comp_units <- model_out_tbl |>
+      split(f = model_out_tbl[, c("model_id", comp_unit_cols)])
+    model_out_tbl <- split_comp_units |>
+      purrr::map(.f = function(split_outputs) {
+        current_comp_unit <- split_outputs |>
+          dplyr::distinct(dplyr::across(dplyr::all_of(comp_unit_cols)), .keep_all = TRUE) |>
```

Comment:
It gives us a way to get/store all the information about the task ID column values for the current iteration so that we can filter out the correct row from the `samples_per_combo` data frame, which gives us how many samples (labeled by unique indices) we're interested in drawing from the current `split_outputs` data frame.

The `keep_all` argument is just a logical asking whether to keep every column when performing the filter for distinct values.